modules/gapi/src/streaming/onevpl/demux/async_mfp_demux_data_provider.cpp

```diff
@@ -0,0 +1,783 @@
+                                                      ComPtrGuard<IMFMediaBuffer> &&buffer,
+                                                      std::shared_ptr<mfxBitstream> &&staging_stream) {
+    size_t bitstream_count = 0;
+    size_t worker_buffer_count = 0;
+    {
+        std::unique_lock<std::mutex> l(buffer_storage_mutex);
+
+        // remember sample buffer to keep data safe
+        worker_key_to_buffer_mapping_storage.emplace(key, std::move(buffer));
+        worker_buffer_count = worker_key_to_buffer_mapping_storage.size();
+
+        // remember bitstream for consuming
+        worker_locked_buffer_storage.push(std::move(staging_stream));
+        bitstream_count = worker_locked_buffer_storage.size();
+        buffer_storage_non_empty_cond.notify_all();
+    }
+    GAPI_DbgAssert(worker_buffer_count == bitstream_count &&
+                   "worker_key_to_buffer_mapping_storage & worker_locked_buffer_storage"
+                   " must be the same size" );
+    GAPI_LOG_DEBUG(nullptr, "[" << this << "] created dmux buffer by key: " <<
+                            key << ", ready bitstream count: " <<
+                            bitstream_count);
+
+    return worker_buffer_count;
+}
+
+/////////////// IDataProvider methods ///////////////
+IDataProvider::mfx_codec_id_type MFPAsyncDemuxDataProvider::get_mfx_codec_id() const {
+    return codec;
+}
+
+mfxStatus MFPAsyncDemuxDataProvider::fetch_bitstream_data(std::shared_ptr<mfxBitstream> &out_bitsream) {
+    if (empty()) {
+        GAPI_LOG_DEBUG(nullptr, "[" << this << "] empty");
+        return MFX_ERR_MORE_DATA;
+    }
+    do {
+        if (out_bitsream) {
+            // make dmux buffer unlock for not empty bitstream
+            GAPI_LOG_DEBUG(nullptr, "[" << this << "] " <<
+                                    "bitstream before fetch: " <<
+                                    out_bitsream.get() <<
+                                    ", DataOffset: " <<
+                                    out_bitsream->DataOffset <<
+                                    ", DataLength: " <<
+                                    out_bitsream->DataLength);
+            if (out_bitsream->DataOffset < out_bitsream->DataLength) {
+                return MFX_ERR_NONE;
+            }
+
+            // cleanup
+            auto it = processing_key_to_buffer_mapping_storage.find(out_bitsream->Data);
+            if (it == processing_key_to_buffer_mapping_storage.end()) {
+                GAPI_LOG_WARNING(nullptr, "[" << this << "] " <<
+                                          "cannot find appropriate dmux buffer by key: " <<
+                                          static_cast<void*>(out_bitsream->Data));
+                GAPI_Assert(false && "invalid bitstream key");
+            }
+            if (it->second) {
+                it->second->Unlock();
+            }
+            processing_key_to_buffer_mapping_storage.erase(it);
```

Comment:
Why this container is not guarded btw?

Reply:
There are two kind of containers: **work_*** and **processing_*** . **Work**-kind of container is filling in `OnReadSample` callback which may be called from different threads. But `fetch_bitsream`  is supposed to be called from single one thread. When application request buffers for decode then it calls `fetch_bitsream` which just make atomic swap of **work_*** and **processing_*** queues. So all **processing_*** data structures  are used from `fetch_bitsrream` only and supposed to be used only from single thread(now)